### DISCUSSION TITLE
How to override autoloaded classes ... and extend them?

### DISCUSSION THREAD

**Discussion by lat9** [zencart/zencart#6542] [General]
Let's say I want (for whatever reason) to override the `Products.php` class.  Since that class is autoloaded via `$psr4Autoload`, I can _replace_ the file by copying it to `includes/classes/MyProduct.php` and adding an init_include containing
```php
<?php
$psr4Autoloader->setClassFile('Product', DIR_FS_CATALOG . DIR_WS_CLASSES . 'MyProduct.php');
```

Now, what hoops need to be jumped through to _extend_ the original class?

**Comment by drbyte** [comment]
Without a "registry" of singleton classes it gets complicated.
It's actually even more complicated when you _replace_ it as you indicated, because now the name of the original class can't be autoresolved.

For `MyProduct.php` to be defined with `extends SomeOtherClass` one would normally add a header in the new class file like `use Product as OriginalProduct` and then the new class gets defined with `... extends OriginalProduct`.
But when you `setClassFile` to load the new one instead of the original, that `use` header won't know what file to load for the original.  You'd have to also `setClassFile('OriginalProduct', ... 'Product.php')` first. 

But we also have Laravel to leverage, which has a registry that can be used. So that'd be a better route to take ... but it means pushing all our calls to said class into the Laravel infrastructure ... which is ultimately planned, but is an invasive code change.
